### PR TITLE
Add Machine Translation API with Microsoft API engine implementation

### PR DESCRIPTION
**Pull request by garionpin**
New MT API added with NoOp and Microsoft API implementation.
Has source leveraging enabled by default with Spring caching enabled.

### CODE REVIEW

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+                    }
+                }
+            } else {
+                logger.debug("No TMTextUnit Match found for content: {}", sourceText);
+            }
+        }
+        return translationsResponseDTO;
+    }
+
+    @SuppressWarnings("SpringCacheableComponentsInspection")
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java:157]: why is that needed?

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    @Cacheable("machineTranslation")
+    List<TextUnitDTO> getLeveragedTextUnitDTOS(String sourceText) {
+        TMTextUnit searchTmTextUnit = new TMTextUnit();
+        searchTmTextUnit.setContent(sourceText);
+
+        return leveragerByContent.getLeveragingMatches(searchTmTextUnit, null, null);
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java:163]: is that fine in term of performance?

Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java:163]: Performance isn't great as that API returns more results than desired generally and the query isn't solely optimized for this purpose, which is why I've included a skipLeveraging parameter - though it is usable if it's not on a critical path - and especially fast on subsequent calls with the in-memory caching being enabled on getLeveragedTextUnitDTOS().

Ideally we'd do this operation in bulk for multiple sources and only return the latest/best match per language, as per the comment from line 122.

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            throw e;
+        }
+    }
+
+    private TranslationsResponse getTranslationsResponseDTO(
+            List<String> textSources, MicrosoftTextTranslationDTO[] response, List<String> targetLocaleTags) {
+        List<TextUnitTranslationGroup> textUnitTranslationGroups = new ArrayList<>();
+
+        // Translation objects in the output for each source string are returned in the same order as provided by the client.
+        for (int i = 0; i < textSources.size(); i++) {
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java:148]: nit: can probably do soemthing like Streams.zip(textSources.stream(), response.stream(), (ts, mt) -> ... ).collect(toList())

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        // Within a set of translations for a specific source string, the languages are ordered in the same way as
+        // the order of target languages provided to the API.
+        for (int j = 0; j < targetLocaleTags.size(); j++) {
+            MicrosoftLanguageTranslationDTO microsoftLanguageTranslationDTO = microsoftTextTranslationDTO.getTranslations().get(j);
+            String requestTargetLocale = targetLocaleTags.get(j);
+
+            Translation translationDTO = new Translation();
+            translationDTO.setTranslationSource(getSource());
+            translationDTO.setLocaleTag(requestTargetLocale);
+            translationDTO.setText(microsoftLanguageTranslationDTO.getText());
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java:172]: nit: I see potential stream everywhere now 😭

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    public static final String OCP_APIM_SUBSCRIPTION_REGION = "Ocp-Apim-Subscription-Region";
+    public static final String X_CLIENT_TRACE_ID = "X-ClientTraceId";
+
+    static Logger logger = LoggerFactory.getLogger(MicrosoftMTEngine.class);
+
+    @Autowired
+    private RestTemplate restTemplate;
+
+    @Autowired
+    private MicrosoftMTEngineConfiguration mtEngineConfiguration;
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java:47]: Auto-wiring the config works and I know the code use that pattern in some place but in that case it might be preferable to keep your service agnostic of spring. 

MicrosoftMTEngine (like S3BlobStorage) seems to be a good candidate to work without spring and be plain Java. 

You'd pass required config via the constructor instead (see S3BlobStorageConfigurationConfiguration). 

Also you end with a central place for setting up the MT engine like (see BlobStorageConfiguration). It is easier to see what the app is doing overall and you get a bit more flexibility (see DatabaseBlobStorageConfiguration it create the blobStorage but also setup job for clean up, etc).

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java (hunk 1 above)
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java:157]: Due to the caching the results of a non-public method. The error is emitted as that would not generally work - except in our case we use AspectJ - which does ensure that functionality is well covered. (Might make that method public at some point though as there might be some value in it being reused like that).

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java (hunk 2 above)
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java:163]: I'll add repositoryId as an additional parameter to be able to narrow the search - if not specified, we'll search in the entire DB. 

Anything that narrows the scope of the search is welcome as far as I'm concerned.

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+                .getTextUnitTranslations().get(0)
+                .getTranslations().get(0);
+    }
+
+    @SuppressWarnings("SpringCacheableComponentsInspection")
+    @Cacheable("machineTranslation")
+    @Timed("MachineTranslationService.getMachineTranslatedResponse")
+    TranslationsResponse getMachineTranslatedResponse(String sourceLocaleTag, List<String> targetLocaleTags, List<String> textSources, Boolean skipFunctionalProtection) {
+        if (textSources.isEmpty()) {
+            return new TranslationsResponse();
```
[the remark this answers is not in the document]
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java:90]: It's to ensure we do not try to ask our specific MT engine implementations to translate empty/null strings - which in some cases results in errors - this ensured a well formatted response in those cases.

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java (hunk 4 above)
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java:172]: We can't use streams here (as with some of the other scenarios) as we need to iterate through two data sources at the same time.
The API response from the MT engine gives us back results in a specific order but with their own language tags - and we want to reply back to our users with the original language tags that they gave us for consistency.

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java (hunk 3 above)
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java:148]: 'zip(java.util.stream.Stream<A>, java.util.stream.Stream<B>, java.util.function.BiFunction<? super A,? super B,R>)' is marked unstable with @Beta 

- are we still Ok to use it given that warning?

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/NoOpEngine.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+public class NoOpEngine implements MachineTranslationEngine{
+    static Logger logger = LoggerFactory.getLogger(NoOpEngine.class);
+
+    @Override
+    public TranslationSource getSource() {
+        return TranslationSource.NOOP;
+    }
+
+    @Override
+    public TranslationsResponse translate(List<String> textSources, String sourceLocaleTag, List<String> targetLocaleTags, String sourceMimeType, String customModel) {
```
[the remark this answers is not in the document]
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/NoOpEngine.java:21]: That's precisely the reason. 

To get the best performance - we should be able to provide a list of sources x a list of locales. If some engines don't support that, the specific engine implementations can then break it down in different types of batches (e.g.: if they support multiple sources for one locale or one source for multiple locales).

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java (hunk 3 above)
Comment (aurambaj) [reply webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java:148]: I'd probably fine. looks like the beta was added later, like this old doc doesn't have it: https://guava.dev/releases/23.0/api/docs/com/google/common/collect/Streams.html

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java (hunk 5 above)
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java]: Done

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngineConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * Configuration settings for the Microsoft Machine Translation client.
+ * Only used when "l10n.mt.impl" has the "MicrosoftMTEngine" value and the
+ * subscription key is provided in the "l10n.microsoft-mt.subscriptionKey" property.
+ *
+ * @author garion
+ */
+@Configuration
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngineConfiguration.java:14]: Not sure about that one, don't you want `@Component @ConfigurationProperties` and no `@ConditionalOnProperty`?

Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngineConfiguration.java:14]: Changed it to the advised setup now.
 
We do use @Configuration together with @ConfigurationProperties in other places: ActuatorHealthLegacyConfig, FlyWayConfig, QuartzPropertiesConfig, DropScheduleConfig, SlaCheckerEmailConfig, SmartlingClientConfiguration & a set of test classes.

The reason why I added @ConditionalOnProperty is to fail early if the engine isn't configured correctly, i.e.: the subscription key is missing - otherwise we'd only fail at runtime, when we actually do the http call.

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+        // TODO(garion): Implement functional protection / placeholder processing
+        if (skipFunctionalProtection != null && skipFunctionalProtection) {
+            logger.debug("function / placeholder protection");
+        }
+
+        return machineTranslationEngine.getTranslationsBySourceText(
+                textSources,
+                sourceBcp47Tag,
+                targetBcp47Tags,
+                "",
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java:130]: empty strings? should that be null, Optional or some enum?

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+                    return new AbstractMap.SimpleEntry<>(sourceText, translations);
+                })
+                .filter(entry -> !entry.getValue().isEmpty())
+                .collect(ImmutableMap.toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private ImmutableMap<String, TextUnitDTO> getMostRecentTranslationsByLanguage(ArrayList<Long> repositoryIds, ArrayList<String> repositoryNames, String sourceText) {
+        List<TextUnitDTO> textUnitDTOsForLeveraging = getLeveragedTextUnitDTOS(sourceText, repositoryIds, repositoryNames);
+
+        Comparator<TextUnitDTO> tmTextUnitCreatedDateComparator = Comparator.comparing(TextUnitDTO::getTmTextUnitCreatedDate);
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java:183]: nit: fine inlining?

File: webapp/src/main/java/com/box/l10n/mojito/service/leveraging/LeveragerByContent.java [hunk 11]
```diff
@@ -40,6 +42,23 @@ public List<TextUnitDTO> getLeveragingMatches(TMTextUnit tmTextUnit, Long source
         return textUnitSearcher.search(textUnitSearcherParameters);
     }
 
+    public List<TextUnitDTO> getLeveragingMatchesForRepository(TMTextUnit tmTextUnit, Long sourceTmId, ArrayList<Long> repositoryIds, ArrayList<String> repositoryNames) {
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/leveraging/LeveragerByContent.java:45]: nit: this should be in another class since it is not used by this leverager. probably fine in the caller class?

**Review by aurambaj — COMMENTED** [review]
looks great! a few question and mostly cosmetic comments

File: webapp/src/main/java/com/box/l10n/mojito/rest/machinetranslation/BatchTranslationRequestDTO.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+ *
+ * @author garion
+ */
+public class BatchTranslationRequestDTO {
+    List<String> textSources;
+    String sourceBcp47Tag;
+    ArrayList<String> targetBcp47Tags;
+    boolean skipFunctionalProtection;
+    boolean skipLeveraging;
+    ArrayList<Long> repositoryIds;
```
[the remark this answers is not in the document]
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/rest/machinetranslation/BatchTranslationRequestDTO.java:17]: Left over from when I had those as params (modeled after TextUnitWS).
Fixed now.

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+        ImmutableMap<String, ImmutableList<TranslationDTO>> machineTranslationsBySourceText = getMachineTranslationBySourceText(
+                getSourceBcp47TagOrDefault(sourceBcp47Tag),
+                targetBcp47Tags,
+                untranslatedStrings, skipFunctionalProtection);
+
+        translationsBySourceText = translationMerger.getMergedTranslationsBySourceText(translationsBySourceText, machineTranslationsBySourceText);
+        return getSortedTranslationsResponse(textSources, targetBcp47Tags, translationsBySourceText);
+    }
+
+    public TranslationDTO getSingleTranslation(
```
[the remark this answers is not in the document]
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java:94]: That is correct - TranslationDTO is meant for the simpler getSingleTranslation() API while TranslationResponseDTO is the fully featured batch response - that also includes the source strings and all of the data in a properly sorted order.

I'd expect most applications, at least initially, to use the simpler variant.

File: webapp/src/main/java/com/box/l10n/mojito/rest/machinetranslation/TranslationRequestDTO.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.util.ArrayList;
+
+/**
+ * DTO to translate one source string for one target language.
+ *
+ * @author garion
+ */
+public class TranslationRequestDTO {
+    String textSource;
```
[the remark this answers is not in the document]
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/rest/machinetranslation/TranslationRequestDTO.java:11]: There are a couple of differences in the parameters expected as you mentioned, but also, the return type for the single translation API method (which we expect initially to account for the bulk of the integrations) is significantly easier to consume than the batch one.

I don't see much benefit in having both a sync and an async variant for exactly the same method and input object and response.

Given the choice - I'd expect the consumers to always choose the simpler, sync, variant to the detriment of the async one. The main reason to have async for the batch is better management of resources on the server side - which I wouldn't expect the clients to care greatly about.

A natural way of achieving good performance and scaling in this case is to keep the small/fast API calls synchronous, while forcing the expensive computations/operations to an async model. 
Doing large batches synchronously isn't great for the server, while doing async (+ the PollableTask overhead) on a call that returns very fast isn't justified either. 
Not exposing those suboptimal choices ensures that the client can't make the wrong choice.

File: webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java (hunk 9 above)
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java:130]: That should be null for now. Once a source placeholder escaping mechanism is introduced, that will be replaced with a relevant enum.

**Review by aurambaj — APPROVED** [review]
